### PR TITLE
Code Llama Fine-tuning Support

### PR DESCRIPTION
**Pull request by adam-weinberger**
@JegernOUTT 
This pull request adds support for fine-tuning Code Llama 7b. The biggest update is in the supported_models.py file which adds the configurations for Code Llama. The second biggest update adds support for additional model configurations (in finetune_train_default.py and finetune_train.py). The other updates are minor and are mostly for logging.

To test this update from command line:

1. Update the finetune.cfg file:
```
{
    "model_name": "codellama/7b",
    "use_heuristics": false,
    "train_steps": 75,
    "batch_size": 16,
    "save_every": 1
}
```
2. Put your code files (easiest is a .zip of the repo) in the folder: ~/.refact/perm-storage/uploaded-files
3. Run: 
```
DEBUG=1 python -m refact.refact_data_pipeline.finetune.process_uploaded_files
DEBUG=1 python -m refact.refact_data_pipeline.finetune.finetune_filter
DEBUG=1 python -m refact.refact_data_pipeline.finetune.finetune_train
```
4. Verify the model has trained correctly by checking the "status" is "finished" in .refact/perm-storage/loras/lora-YYYYmmDD-MMSSss/status.json

To test this update from the UI:

1. Run:
```
DEBUG=1 python -m self_hosting_machinery.webgui.webgui
DEBUG=1 python -m self_hosting_machinery.inference.inference_worker --model codellama/7b
```
2. On the Model Hosting select "codellama/7b". Click through the "Sources" and "Finetune" tabs as normal, being sure To select the "codellama/7b" model under "Start New Fine Tune". NOTE: sometimes portions of the UI are grayed out when I try to go through this workflow and I'm not always able to completed. Is this expected? I have not touched any of the UI code.
3. Verify how the model has trained correctly by looking in the "Completed Runs"

### CODE REVIEW

File: refact_data_pipeline/finetune/supported_models.py [hunk 1]
```diff
@@ -151,5 +151,43 @@
...
+            "backproj": ["attn.c_proj"],
+            "mlp": ["mlp.c_fc", "mlp.c_proj"],
+        },
+        "freeze_exceptions_mapping": {
+            "wte": "wte",
+            "lm_head": "lm_head",
+            "lora": "lora"
+        },
+        "tokenizer": {
+            "eot_idx": 0,
```
Comment (JegernOUTT) [inline refact_data_pipeline/finetune/supported_models.py:169]: I think there is a problem since those indexes are for `bigcode` encoding, and you are using a llama encoding

They also use slightly different encoding format (with `bos` token) and maybe the `FIM` format is also different

**Review by JegernOUTT — CHANGES_REQUESTED** [review]

File: refact_data_pipeline/finetune/supported_models.py (hunk 1 above)
Comment (adam-weinberger) [reply refact_data_pipeline/finetune/supported_models.py:169]: Sorry, just saw this. I understand what you are saying, but I'm a little unclear on how to find the correct encoding. This is the [Code Llama tokenizer](https://huggingface.co/docs/transformers/main/model_doc/code_llama#transformers.CodeLlamaTokenizer). Should I create a new `code_llama.json` In the refact_encoding folder with these tokens? Where did the `bigcode_largemodel.json` file come from? Where did the vocab part of the file come from? When I google the entire file for Code Llama the best result is [this](https://huggingface.co/MarioPenguin/beto_amazon/commit/83bd1c1ba9422bc57e90b92f93b4bd112afea8be.diff?file=tokenizer.json) which is for BERT.

Comment (JegernOUTT) [reply refact_data_pipeline/finetune/supported_models.py:169]: We are using encodings which are downloaded from HF directly. Since you're using https://huggingface.co/TheBloke/CodeLlama-7B-fp16, try to study the [tokenizer file](https://huggingface.co/TheBloke/CodeLlama-7B-fp16/raw/main/tokenizer.json), [code llama FIM format](https://github.com/ggerganov/llama.cpp/issues/2818). Compare them with some official sources. 
As an `escape` token you can choose any "empty" token, treat it like some kind of "control" symbol

Also, don't forget about `BOS` token, as I remember llama models are using it as a starting label. If it's true - make sure our datapipeline prepares data correctly

Comment (adam-weinberger) [reply refact_data_pipeline/finetune/supported_models.py:169]: Yes, that makes sense. I will look into it.

Alternatively, you are sure that lines 135-143 in encoding.py, are not a valid method of applying the llama tokenization?

```
        elif name in ['llama']:
            from sentencepiece import SentencePieceProcessor
            filename = Path(__file__).resolve().parent / f"{name}.tokenizer.model"
            self._sentencepiece_tokenizer = SentencePieceProcessor(str(filename))
            self.n_vocab = self._sentencepiece_tokenizer.vocab_size()
            self.bos_id: int = self._sentencepiece_tokenizer.bos_id()
            self.DIAMOND = self._sentencepiece_tokenizer.unk_id()
            self.EOT = self._sentencepiece_tokenizer.eos_id()
            self.LF = 13
```

Comment (JegernOUTT) [reply refact_data_pipeline/finetune/supported_models.py:169]: It should be valid. 
However, we don't use this code unless you're using an old `codify` model
See https://github.com/smallcloudai/refact/blob/main/refact_data_pipeline/finetune/model_handling.py#L127 to get the idea


I must mention, that the logic behind constructing all these objects is weird, that's why we are refactoring it right now (https://github.com/smallcloudai/refact/pull/182)

**Comment by JegernOUTT** [comment]
@adam-weinberger please, rebase your branch onto https://github.com/smallcloudai/refact/tree/v1.2.0 and resolve all conflicts

**Review by JegernOUTT — APPROVED** [review]
